doc/pattern.md

```diff
@@ -0,0 +1,439 @@
+
+
+The method `statement` below defines a Spoon pattern. 
+
+```java
+public class Foo {
+	public void statement() {
+		if (_col_.size() > 10)
+			throw new OutOfBoundException();
+	}
+}
+```
+
+The code, which creates a Spoon pattern using `PatternBuilder` looks like this:
+
+```java
+Pattern t = PatternBuilder.create(factory,
+	//defines pattern class.
+	CheckBoundTemplate.class, 
+	//defines which part of pattern class will be used as pattern model
+	model -> model.setBodyOfMethod("statement"))
+	//tells builder that all variables defined out of scope of pattern model (body of the method)
+	//are considered as pattern parameters
+	// here _col_
+	.configureTemplateParameters()
+	//builds an instance of Pattern
+	.build();
+```
+
+
+This pattern specifies a
+statements (all statements of body of method `statement`) that is a precondition to check that a list 
+is smaller than a certain size. This piece of code will be injected at the 
+beginning of all methods dealing with size-bounded lists. This pattern has 
+one single pattern parameter called `_col_`. 
+In this case, the pattern parameter value is meant to be an expression (`CtExpression`) 
+that returns a Collection.
+
+The pattern source is 
+well-typed, compiles, but the binary code of the pattern is usually thrown away
+and only spoon model (Abstract syntax tree) of source code is used to generated new code
+or to search for matching code.
+
+Generating of code using Spoon pattern
+-------------
+
+The code at the end of this page shows how to use such spoon pattern.
+One takes a spoon pattern, defines the pattern parameters, 
+and then one calls the pattern engine. In last line, the bound check 
+is injected at the beginning of a method body. 
+
+Since the pattern is given the first method parameter which is in the 
+scope of the insertion location, the generated code is guaranteed to compile. 
+The Java compiler ensures that the pattern compiles with a given scope, the 
+developer is responsible for checking that the scope where she uses 
+pattern-generated code is consistent with the pattern scope.
+
+```java
+Pattern t = PatternBuilder.create(...building a pattern. See code above...);
+// creating a holder of parameters
+Map<String, Object> parameters = new HashMap<>();
+parameters.put("_col_", createVariableAccess(method.getParameters().get(0))); 
+
+// getting the final AST
+CtStatement injectedCode = t.substituteSingle(factory, CtStatement.class, parameters);
+
+// adds the bound check at the beginning of a method
+method.getBody().insertBegin(injectedCode);
+```
+
+## PatternBuilder parameters
+The `PatternBuilder` takes all the Template parameters mentioned in the chapters above
+and understands them as pattern parameters, when `PatternBuilder#configureTemplateParameters()`
+is called.
+
+```java
+Pattern t = PatternBuilder.create(...select pattern model...)
+	.configureTemplateParameters()
+	.build();
+```
```

Comment:
The description from `## PatternBuilder parameters` till here is related to integration of legacy templates with new Pattern. It probably doesn't belong into this document ... or is at lest confusing in current content.